R/bootstrap.R

```diff
@@ -757,18 +777,39 @@ tabsetPanel <- function(...,
   if (!is.null(id))
     selected <- restoreInput(id = id, default = selected)
 
-  # build the tabset
-  tabs <- list(...)
   type <- match.arg(type)
+  tabset <- buildTabset(..., ulClass = paste0("nav nav-", type), id = id, selected = selected)
+
+  nav <- tabset$navList
+  if (card) {
+    nav <- tags$div(
+      class = "card-header",
+      tagAppendAttributes(
+        nav, class = paste0("card-header-", type)
+      ),
+      tagFunction(function() {
+        if (getCurrentVersion() >= 4) {
+          return(NULL)
+        }
+        warning(
+          "`tabsetPanel(card = TRUE)` requires Bootstrap 4 or higher, ",
+          "so the app has been upgraded from Bootstrap 3 to 4. ",
+          "To remove this warning, either supply `theme = bslib::bs_theme()` ",
+          "to the app's page layout or set `card = FALSE`.",
+          call. = FALSE
+        )
+        bootstrapLib(bslib::bs_theme(version = 4))
```

Comment:
This is probably a little too magical -- any tags that would appear above this `tagFunction()` would end up generating BS3 markup....